lgc/util/ElfNoteEntryInsertionUtil.cpp

```diff
@@ -0,0 +1,258 @@
+// In particular, see "Notes (Nhdr)" for the document of the note section.
+
+namespace {
+
+// It is similar to struct NoteHeader defined in llpc/util/vkgcElfReader.h, but
+// it allows us to use Elf_Nhdr_Impl<object::ELF64LE>::Align and it does not
+// have the size limitation of note name.
+using NoteHeader = object::Elf_Nhdr_Impl<object::ELF64LE>;
+
+// An array of zeros. We use it for the note alignment.
+constexpr char ZerosForNoteAlign[NoteHeader::Align] = {'\0'};
+
+// =====================================================================================================================
+// Contents of a section to be shifted and its new offset.
+struct SectionShiftInfo {
+  SmallString<64> section;
+  ELF::Elf64_Off newOffset;
+};
+
+// =====================================================================================================================
+// Add an entry to the note section.
+//
+// Reference: Linux Programmer's Manual ELF(5) "Notes (Nhdr)".
+//
+// @param noteName              : Name of the note entry.
+// @param noteDesc              : Description of the note entry.
+// @param [out] noteEntryWriter : The .note section where the note entry will be added to.
+void addNoteEntry(StringRef noteName, ArrayRef<uint8_t> noteDesc, BinaryStreamWriter &noteEntryWriter) {
+  NoteHeader noteHeader = {};
+  noteHeader.n_namesz = noteName.size() + 1;
+  noteHeader.n_descsz = noteDesc.size();
+  // TODO: Define a note type to specify cache hash and llpc version. It must
+  // be defined in llvm/include/llvm/BinaryFormat/ELF.h.
+  // Note types with values between 0 and 32 (inclusive) are reserved.
+  noteHeader.n_type = 0;
+  auto error = noteEntryWriter.writeObject(noteHeader);
+  (void)error;
+  assert(!error);
+
+  // Write the note name terminated by zero and zeros for the alignment.
+  error = noteEntryWriter.writeCString(noteName);
+  assert(!error);
+  error = noteEntryWriter.writeFixedString(StringRef(
+      ZerosForNoteAlign, offsetToAlignment(noteEntryWriter.getLength(), Align::Constant<NoteHeader::Align>())));
+  assert(!error);
+
+  // Write the note description and zeros for the alignment.
+  error = noteEntryWriter.writeBytes(noteDesc);
+  assert(!error);
+  error = noteEntryWriter.writeFixedString(StringRef(
+      ZerosForNoteAlign, offsetToAlignment(noteEntryWriter.getLength(), Align::Constant<NoteHeader::Align>())));
+  assert(!error);
+}
+
+// =====================================================================================================================
+// Prepare a byte-stream for note entries to be added to the note section.
+//
+// @param notes                 : The array of note entries to be added to the note section.
+// @param newNoteEntryOffset    : The offset in ELF where the note entries will be added.
+// @param [out] noteEntryStream : The byte-stream to be filled with the note entries.
+void prepareNoteEntries(ArrayRef<NoteEntry> notes, const ELF::Elf64_Off newNoteEntryOffset,
```

Comment:
What does it mean to prepare note entries here? I can see that this functions adds them in the loop. Can you make the docstring and/or the function name less generic sounding? Would `addNoteEntries` work?